### DISCUSSION TITLE
Examples with None values are not included into `openapi.json`

### DISCUSSION THREAD

**Discussion by mezrin** [fastapi/fastapi#12048] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel

class MyModelClass(BaseModel):
    field_a: int
    field_b: int | None

    model_config = {
        "json_schema_extra": {
            "examples": [
                {"field_a": 123, "field_b": None},
            ]
        }
    }


@router.post("/my_endpoint")
def my_endpoint() -> MyModelClass:
    return MyModelClass(field_a=123, field_b=None)
```

### Description

My endpoint doesn't exclude None/null values. If value is null - then the client of API receives null.

But when I load the swagger UI - example of the output doesn't include `field_b`. It looks like this: `{ "field_a": 123 }`.
When I open `openapi.json` - there is also no `field_b`. I contains this: `"examples":[{"field_a":123}]`

Fastapi' docs state the example is added to the openapi.json as is: `That extra info will be added as-is to the output JSON Schema for that model, and it will be used in the API docs.`
https://fastapi.tiangolo.com/tutorial/schema-extra-example/#extra-json-schema-data-in-pydantic-models
But in reality it's not. Looks like a bug. 

### Operating System

Linux

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

3.12.0

**Comment by JavierSanchezCastro** [comment]
Duplicate of https://github.com/fastapi/fastapi/discussions/8401
Still without a solution

**Comment by sinisaos** [comment]
@mezrin You can try this little workaround using `json.dumps` (to convert Python objects to a json array) like this

```python
import json
from fastapi import FastAPI
from pydantic import BaseModel


def converter(data: dict) -> str:
    return json.dumps(data)


class MyModelClass(BaseModel):
    field_a: int
    field_b: int | None

    model_config = {
        "json_schema_extra": {
            "examples": [
                converter({"field_a": 123, "field_b": None}),
            ]
        }
    }


app = FastAPI()


@app.post("/my_endpoint")
def my_endpoint() -> MyModelClass:
    return MyModelClass(field_a=123, field_b=None)
```

The result is this

![example](https://github.com/user-attachments/assets/79bfc0f7-0fea-446f-ac00-4f6bd633d41e)


I hope you find this useful.

**Reply by mezrin** [reply]
thank you for idea!
This converts the examples to string and preserves `null` values.
Swagger UI then properly displays it.

The problem is that we use 3rd party tool to provide access to the API docs (ApiDog). And, unlike Swagger, this tool doesn't parse the example converted to string. So in the example section I see the string with escape characters, not formatted etc. 

Keep looking for the solution.
Seems I'll just make "postprocessing" of openapi.json to add stripped fields 🤦

**Reply by sinisaos** [reply]
@mezrin You are right. `ReDoc` or `ApiDog` do not parse and format as well as `Swagger UI`. You can try this to convert `None` to `"null" as a string` like this
```python
import json

model_config = {
    "json_schema_extra": {
        "examples": [
            {"field_a": 123, "field_b": json.dumps(None)},
        ]
    }
}
# result is 
{
  "field_a": 123,
  "field_b": "null"
}
```
If you are not satisfied with the result, as you say, you can edit `openapi.json` and remove the quotes from `null` and then import `openapi.json` into `ApiDog` and the result would be 
```json
{
    "field_a": 123,
    "field_b": null
}
```
If you want to use ReDoc, simply add a custom `openapi.json` to your FastAPI application like this 

```python
import json

app = FastAPI()

app.openapi_schema = json.load(open("custom_openapi.json"))
```

and the result will be the same as in `ApiDog`

**Reply by mezrin** [reply]
@sinisaos sounds good! thank you for ideas !

**Comment by mezrin** [comment]
After lots of struggles I did the following:
- prepared the examples as I need them.
- serialised examples into strings and added to the models: `model_config = { "json_schema_extra": { "examples": my_examples } }`
- created a script for 50 lines that reads `openapi.json`, finds all `"examples": [str...]`, converts each example from str to a json object, persists updated content of `openapi.json` back to the file.

Now all works fine for me. SwaggerUI parses examples serialised to strings - you need to do nothing to make Swagger work. When i export data to ApiDog - i use script to process `openapi.json` and after that ApiDog displays all correctly as well.

---

also I found out that with this approach it's not required to set model_config statically. Instead you can do it later like this: `MyModel.model_config = {"json_schema_extra": {"examples": my_examples}}`
This allows to use the model itself to prepare examples: create an instance of the model, fill with data, serialise with `my_instance.model_dump_json()`

**Comment by YuriiMotov** [comment]
Let's close this topic and track it in : https://github.com/fastapi/fastapi/discussions/8401
